infrastructure/src/main/java/org/corfudb/infrastructure/management/failuredetector/DecisionMakerAgent.java

```diff
@@ -0,0 +1,113 @@
+package org.corfudb.infrastructure.management.failuredetector;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.management.ClusterAdvisor;
+import org.corfudb.protocols.wireprotocol.ClusterState;
+import org.corfudb.protocols.wireprotocol.NodeState;
+import org.corfudb.protocols.wireprotocol.failuredetector.FileSystemStats;
+import org.corfudb.protocols.wireprotocol.failuredetector.FileSystemStats.PartitionAttributeStats;
+import org.corfudb.protocols.wireprotocol.failuredetector.FileSystemStats.ResourceQuotaStats;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank.NodeRankByPartitionAttributes;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank.NodeRankByResourceQuota;
+
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.TreeSet;
+
+@AllArgsConstructor
+@Slf4j
+public class DecisionMakerAgent {
+    private final ClusterState clusterState;
+    private final ClusterAdvisor clusterAdvisor;
+
+    /**
+     * Find a node that is a decision maker according to the current cluster state
+     *
+     * @return a decision maker
+     */
+    public Optional<String> findDecisionMaker() {
+        if (!clusterState.getLocalNode().isPresent()) {
+            return Optional.empty();
+        }
+
+        Optional<String> partitionDm = findPartitionAttributesDecisionMaker()
+                .map(NodeRankByPartitionAttributes::getEndpoint);
+
+        Optional<String> quotaDm = findResourceQuotaDecisionMaker()
+                .map(NodeRankByResourceQuota::getEndpoint);
+
+        Optional<String> clusterDm = clusterAdvisor.findDecisionMaker(clusterState)
+                .map(NodeRank::getEndpoint);
+
+        //Only completely healthy nodes can update cluster layout
+        if (clusterDm.equals(partitionDm) && clusterDm.equals(quotaDm)) {
+            return clusterDm;
+        }
+
+        log.trace("Decision maker not found");
+        return Optional.empty();
+    }
+
+    @VisibleForTesting
+    Optional<NodeRankByResourceQuota> findResourceQuotaDecisionMaker() {
+        NavigableSet<NodeRankByResourceQuota> set = new TreeSet<>();
+
+        for (NodeState node : clusterState.getNodes().values()) {
+            String nodeEndpoint = node.getConnectivity().getEndpoint();
+            node.getFileSystem()
+                    .map(FileSystemStats::getResourceQuotaStats)
+                    //nodes with exceeded quota can't be decision makers
+                    .filter(ResourceQuotaStats::isNotExceeded)
+                    .map(quota -> new NodeRankByResourceQuota(nodeEndpoint, quota))
+                    .ifPresent(set::add);
+        }
+
+        Optional<NodeRankByResourceQuota> maybeDecisionMaker = Optional.ofNullable(set.pollFirst());
+
+        return maybeDecisionMaker
```

Comment:
This piece of code is the same for both findPartitionAttributesDecisionMaker and findResourceQuotaDecisionMaker, move it to a separate function?

Reply:
You can also create a function findPartitionAttributeBy and parameterize it with lambda predicate

Reply:
Yeah, I should do that, anyway I just deleted ResourceQuota functionality, so there is no code duplication anymore.